File: clients/ios/HermesStudio/Theme/BoringAvatar.swift
import SwiftUI

/// A Swift port of `boring-avatars-vanilla` (MIT), the generator the web client
/// uses in `packages/client/src/components/hermes/profiles/ProfileAvatar.vue`.
///
/// Only the `beam` variant is ported, because that is the only one the web
/// asks for. Every helper below is a literal translation of
/// `node_modules/boring-avatars-vanilla/dist/index.js` so that the same seed
/// yields the same avatar in the browser and on the phone:
///
/// ```js
/// const hashCode = (name) => { let hash = 0
///   for (let i = 0; i < name.length; i++) {
///     const character = name.charCodeAt(i)
///     hash = (hash << 5) - hash + character; hash = hash & hash }
///   return Math.abs(hash) }
/// ```
///
/// `charCodeAt` walks UTF-16 code units (so an emoji contributes two), the
/// intermediate arithmetic wraps at 32 bits, and `Math.abs` of `Int32.min`
/// widens to 2_147_483_648 — all three are reproduced here.
enum BoringAvatar {
    /// The library default, used because `ProfileAvatar.vue` passes no palette.
    static let defaultPalette = ["#92A1C6", "#146A7C", "#F0AB3D", "#C271B4", "#C20D90"]

    /// The beam variant's viewBox, `SIZE` in the library.
    static let size: Double = 36

    static func hash(_ name: String) -> Int {
        var value: Int32 = 0
        for unit in name.utf16 {
            value = (value &<< 5) &- value &+ Int32(unit)
        }
        return abs(Int(value))
    }

    /// `Math.floor(number / 10 ** position % 10)`. `number` is never negative
    /// here, so integer division is exact and matches the JS float maths.
    static func digit(_ number: Int, _ position: Int) -> Int {
        var divisor = 1
        for _ in 0..<max(position, 0) { divisor *= 10 }
        return (number / divisor) % 10
    }

    /// `getBoolean` — true when the digit at `position` is even.
    static func boolean(_ number: Int, _ position: Int) -> Bool {
        digit(number, position) % 2 == 0
    }

    /// `getUnit`. `index` 0 means "no index": JS treats `0` as falsy and never
    /// negates, so the default argument keeps that behaviour.
    static func unit(_ number: Int, _ range: Int, index: Int = 0) -> Int {
        let value = number % range
        if index != 0 && digit(number, index) % 2 == 0 { return -value }
        return value
    }

    static func randomColor(_ number: Int, _ colors: [String]) -> String {
        colors[number % colors.count]
    }

    /// `getContrast` — the YIQ test that picks the face colour.
    static func contrast(_ hex: String) -> String {
        let digits = Array(hex.hasPrefix("#") ? String(hex.dropFirst()) : hex)
        guard digits.count >= 6 else { return "#FFFFFF" }
        func component(_ start: Int) -> Double {
            Double(Int(String(digits[start..<(start + 2)]), radix: 16) ?? 0)
        }
        let yiq = (component(0) * 299 + component(2) * 587 + component(4) * 114) / 1000
        return yiq >= 128 ? "#000000" : "#FFFFFF"
    }

    /// Everything `generateData` produces for the `beam` variant.
    struct Beam: Equatable {
        var wrapperColor: String
        var faceColor: String
        var backgroundColor: String
        var wrapperTranslateX: Double
        var wrapperTranslateY: Double
        var wrapperRotate: Double
        var wrapperScale: Double
        var isMouthOpen: Bool
        var isCircle: Bool
        var eyeSpread: Int
        var mouthSpread: Int
        var faceRotate: Double
        var faceTranslateX: Double
        var faceTranslateY: Double
    }

    static func beam(name: String, colors: [String] = defaultPalette) -> Beam {
        let number = hash(name)
        let wrapperColor = randomColor(number, colors)

        let preTranslateX = unit(number, 10, index: 1)
        let translateX = preTranslateX < 5 ? Double(preTranslateX) + size / 9 : Double(preTranslateX)
        let preTranslateY = unit(number, 10, index: 2)
        let translateY = preTranslateY < 5 ? Double(preTranslateY) + size / 9 : Double(preTranslateY)

        return Beam(
            wrapperColor: wrapperColor,
            faceColor: contrast(wrapperColor),
            backgroundColor: randomColor(number + 13, colors),
            wrapperTranslateX: translateX,
            wrapperTranslateY: translateY,
            wrapperRotate: Double(unit(number, 360)),
            wrapperScale: 1 + Double(unit(number, Int(size / 12))) / 10,
            isMouthOpen: boolean(number, 2),
            isCircle: boolean(number, 1),
            eyeSpread: unit(number, 5),
            mouthSpread: unit(number, 3),
            faceRotate: Double(unit(number, 10, index: 3)),
            faceTranslateX: translateX > size / 6 ? translateX / 2 : Double(unit(number, 8, index: 1)),
            faceTranslateY: translateY > size / 6 ? translateY / 2 : Double(unit(number, 7, index: 2))
        )
    }

    /// `ProfileAvatar.vue`'s seed: `props.avatar?.seed || (props.name || 'default')`.
    static func seed(avatarSeed: String?, name: String) -> String {
        avatarSeed?.nilIfEmpty ?? name.nilIfEmpty ?? "default"
    }

    /// `#RRGGBB` → `Color`. The palette and `getContrast` only ever produce
    /// six-digit hex, so anything else falls back to clear rather than crash.
    static func color(_ hex: String) -> Color {
        let text = hex.hasPrefix("#") ? String(hex.dropFirst()) : hex
        guard text.count == 6, let value = Int(text, radix: 16) else { return .clear }
        return Color(
            red: Double((value >> 16) & 0xFF) / 255,
            green: Double((value >> 8) & 0xFF) / 255,
            blue: Double(value & 0xFF) / 255
        )
    }
}

/// Draws one `beam` avatar with Core Graphics instead of shipping an SVG
/// renderer. The geometry is the library's markup, in its own 36 × 36 user
/// space, scaled to the requested size.
struct BeamAvatarView: View {
    let seed: String
    var size: CGFloat = 40

    private var beam: BoringAvatar.Beam { BoringAvatar.beam(name: seed) }

    var body: some View {
        Canvas(rendersAsynchronously: false) { context, canvasSize in
            let scale = canvasSize.width / CGFloat(BoringAvatar.size)
            draw(in: &context, base: CGAffineTransform(scaleX: scale, y: scale), beam: beam)
        }
        .frame(width: size, height: size)
        .clipShape(Circle())
        .accessibilityHidden(true)
    }

    /// `<rect width=36 height=36 fill=backgroundColor />`, the transformed
    /// wrapper, then the face group. `base` maps the library's 36 × 36 user
    /// space onto the view; every element transform is concatenated onto it.
    private func draw(in context: inout GraphicsContext, base: CGAffineTransform, beam: BoringAvatar.Beam) {
        let box = CGRect(x: 0, y: 0, width: BoringAvatar.size, height: BoringAvatar.size)

        var background = context
        background.transform = base
        background.fill(Path(box), with: .color(BoringAvatar.color(beam.backgroundColor)))

        var wrapper = context
        wrapper.transform = wrapperTransform(beam).concatenating(base)
        // `rx="36"` is clamped by SVG to half the width, i.e. a circle.
        let radius = beam.isCircle ? BoringAvatar.size / 2 : BoringAvatar.size / 6
        wrapper.fill(
            Path(roundedRect: box, cornerRadius: radius),
            with: .color(BoringAvatar.color(beam.wrapperColor))
        )

        var face = context
        face.transform = faceTransform(beam).concatenating(base)
        drawFace(in: &face, beam: beam)
    }

    /// `translate(tx ty) rotate(r 18 18) scale(s)` — SVG applies the list
    /// right to left, so scale runs first.
    private func wrapperTransform(_ beam: BoringAvatar.Beam) -> CGAffineTransform {
        let centre = BoringAvatar.size / 2
        return CGAffineTransform.identity
            .translatedBy(x: beam.wrapperTranslateX, y: beam.wrapperTranslateY)
            .translatedBy(x: centre, y: centre)
            .rotated(by: beam.wrapperRotate * .pi / 180)
            .translatedBy(x: -centre, y: -centre)
            .scaledBy(x: beam.wrapperScale, y: beam.wrapperScale)
    }

    /// `translate(ftx fty) rotate(fr 18 18)`.
    private func faceTransform(_ beam: BoringAvatar.Beam) -> CGAffineTransform {
        let centre = BoringAvatar.size / 2
        return CGAffineTransform.identity
            .translatedBy(x: beam.faceTranslateX, y: beam.faceTranslateY)
            .translatedBy(x: centre, y: centre)
            .rotated(by: beam.faceRotate * .pi / 180)
            .translatedBy(x: -centre, y: -centre)
    }

    private func drawFace(in context: inout GraphicsContext, beam: BoringAvatar.Beam) {
        let ink = GraphicsContext.Shading.color(BoringAvatar.color(beam.faceColor))
        let mouthY = 19.0 + Double(beam.mouthSpread)

        if beam.isMouthOpen {
            // `M15 y c2 1 4 1 6 0`, stroke width 1, round caps.
            var path = Path()
            path.move(to: CGPoint(x: 15, y: mouthY))
            path.addCurve(
                to: CGPoint(x: 21, y: mouthY),
                control1: CGPoint(x: 17, y: mouthY + 1),
                control2: CGPoint(x: 19, y: mouthY + 1)
            )
            context.stroke(path, with: ink, style: StrokeStyle(lineWidth: 1, lineCap: .round))
        } else {
            // `M13,y a1,0.75 0 0,0 10,0` — the radii are too small for the
            // 10-unit chord, so SVG scales them up to 5 × 3.75; sweep 0 puts
            // the bulge below the line. Filling closes the chord. Written as
            // two Bézier quarters rather than `addArc`, whose `clockwise`
            // flag reads differently in a y-down space.
            let k = 0.5522847498307936
            var path = Path()
            path.move(to: CGPoint(x: -1, y: 0))
            path.addCurve(
                to: CGPoint(x: 0, y: 1),
                control1: CGPoint(x: -1, y: k),
                control2: CGPoint(x: -k, y: 1)
            )
            path.addCurve(
                to: CGPoint(x: 1, y: 0),
                control1: CGPoint(x: k, y: 1),
                control2: CGPoint(x: 1, y: k)
            )
            path.closeSubpath()
            let mouth = CGAffineTransform(translationX: 18, y: mouthY).scaledBy(x: 5, y: 3.75)
            context.fill(path.applying(mouth), with: ink)
        }

        // Two 1.5 × 2 rects with `rx="1"`: SVG clamps rx to 0.75 and keeps
        // ry at 1, so each corner radius covers the whole rect — an ellipse.
        for x in [14 - Double(beam.eyeSpread), 20 + Double(beam.eyeSpread)] {
            let eye = CGRect(x: x, y: 14, width: 1.5, height: 2)
            context.fill(Path(ellipseIn: eye), with: ink)
        }
    }
}

File: clients/android/app/src/main/java/us/i3u/hermesstudio/BoringAvatar.kt
package us.i3u.hermesstudio

import kotlin.math.abs

/**
 * The `beam` variant of boring-avatars, ported from the library Core Hub runs
 * in the browser (`node_modules/boring-avatars-vanilla/dist/index.js`, MIT) so
 * that a seed draws the very same avatar on the phone as it does on the web.
 *
 * `packages/client/src/components/hermes/profiles/ProfileAvatar.vue` calls the
 * library with only `name`, `variant: 'beam'` and `size`, so the palette, the
 * `title` flag and the `square` flag all stay at their defaults here too.
 *
 * Everything below mirrors the JavaScript literally, including the whitespace
 * of the emitted markup: `BoringAvatarTest` compares this output against
 * fixtures captured from the library itself.
 */
object BoringAvatar {

    /** The library default palette, which the web never overrides. */
    val DEFAULT_COLORS: List<String> =
        listOf("#92A1C6", "#146A7C", "#F0AB3D", "#C271B4", "#C20D90")

    /** The beam variant draws on a 36×36 canvas. */
    private const val SIZE = 36

    /**
     * The values the variant derives from the seed. Kept as a type of its own
     * because the tests assert on them directly, and because the arithmetic is
     * the part that has to match the web exactly.
     */
    data class Beam(
        val wrapperColor: String,
        val faceColor: String,
        val backgroundColor: String,
        val wrapperTranslateX: Int,
        val wrapperTranslateY: Int,
        val wrapperRotate: Int,
        /** 0, 1 or 2 — the tenths added to a scale of 1. */
        val wrapperScaleTenths: Int,
        val isMouthOpen: Boolean,
        val isCircle: Boolean,
        val eyeSpread: Int,
        val mouthSpread: Int,
        val faceRotate: Int,
        /**
         * Halved in the library by a plain `/ 2`, which in JavaScript keeps the
         * half: an odd wrapper offset gives the face a `.5` of its own.
         */
        val faceTranslateX: Double,
        val faceTranslateY: Double,
    )

    /**
     * `hashCode` in the library: a 32-bit `h * 31 + c` walk over UTF-16 code
     * units, then an absolute value. The absolute value is taken as a Long
     * because JavaScript's `Math.abs` on -2^31 yields 2147483648, which does
     * not fit back into a 32-bit integer.
     */
    fun hash(name: String): Long {
        var value = 0
        for (index in name.indices) {
            val code = name[index].code
            value = (value shl 5) - value + code
        }
        return abs(value.toLong())
    }

    /** `getDigit`: the nth decimal digit, counting from the units. */
    private fun digit(number: Long, position: Int): Int {
        var divisor = 1L
        repeat(position) { divisor *= 10 }
        return ((number / divisor) % 10).toInt()
    }

    /** `getBoolean`: true when that digit is even. */
    private fun bool(number: Long, position: Int): Boolean = digit(number, position) % 2 == 0

    /**
     * `getUnit`: `number % range`, made negative when the digit at `position`
     * is even. `position` 0 means "always positive", matching the library's
     * falsy-index check.
     */
    private fun unit(number: Long, range: Int, position: Int = 0): Int {
        val value = (number % range).toInt()
        return if (position != 0 && digit(number, position) % 2 == 0) -value else value
    }

    /** `getRandomColor`. */
    private fun colorAt(number: Long, colors: List<String>): String =
        colors[(number % colors.size).toInt()]

    /**
     * `getContrast`: black on a light wrapper, white on a dark one, using the
     * same luminance weights the library uses.
     */
    fun contrast(hexColor: String): String {
        val hex = hexColor.removePrefix("#")
        val red = hex.substring(0, 2).toInt(16)
        val green = hex.substring(2, 4).toInt(16)
        val blue = hex.substring(4, 6).toInt(16)
        return if ((red * 299 + green * 587 + blue * 114) / 1000.0 >= 128) "#000000" else "#FFFFFF"
    }

    /** `generateData` for the beam variant. */
    fun data(name: String, colors: List<String> = DEFAULT_COLORS): Beam {
        val number = hash(name)
        val wrapperColor = colorAt(number, colors)

        val rawX = unit(number, 10, 1)
        val translateX = if (rawX < 5) rawX + SIZE / 9 else rawX
        val rawY = unit(number, 10, 2)
        val translateY = if (rawY < 5) rawY + SIZE / 9 else rawY

        return Beam(
            wrapperColor = wrapperColor,
            faceColor = contrast(wrapperColor),
            backgroundColor = colorAt(number + 13, colors),
            wrapperTranslateX = translateX,
            wrapperTranslateY = translateY,
            wrapperRotate = unit(number, 360),
            wrapperScaleTenths = unit(number, SIZE / 12),
            isMouthOpen = bool(number, 2),
            isCircle = bool(number, 1),
            eyeSpread = unit(number, 5),
            mouthSpread = unit(number, 3),
            faceRotate = unit(number, 10, 3),
            faceTranslateX = if (translateX > SIZE / 6) translateX / 2.0 else unit(number, 8, 1).toDouble(),
            faceTranslateY = if (translateY > SIZE / 6) translateY / 2.0 else unit(number, 7, 2).toDouble(),
        )
    }

    /**
     * JavaScript prints `1` for a whole number and `1.2` otherwise; the scale
     * is always 1 plus nought, one or two tenths, so this covers every case.
     */
    private fun scaleText(tenths: Int): String = if (tenths == 0) "1" else "1.$tenths"

    /**
     * JavaScript drops a trailing `.0` — and prints `-0` as `0` — when it turns
     * a number into text, which is what the transforms rely on.
     */
    private fun numberText(value: Double): String =
        if (value == Math.floor(value)) value.toLong().toString()
        else String.format(java.util.Locale.ROOT, "%s", value)

    /**
     * The mask needs an id that is unique inside one document. The library
     * counts up and appends a timestamp; a seed-derived id is stable instead,
     * which keeps a redrawn avatar byte-identical to the cached one.
     */
    fun idFor(name: String): String = "boring-avatar-${hash(name)}"

    /** The SVG the web renders for this seed, whitespace included. */
    fun beam(
        name: String,
        size: Int = 40,
        colors: List<String> = DEFAULT_COLORS,
        id: String = idFor(name),
    ): String {
        val data = data(name, colors)
        val half = SIZE / 2
        val mouth = if (data.isMouthOpen) {
            """<path d="M15 ${19 + data.mouthSpread}c2 1 4 1 6 0" stroke="${data.faceColor}" fill="none" stroke-linecap="round" />"""
        } else {
            """<path d="M13,${19 + data.mouthSpread} a1,0.75 0 0,0 10,0" fill="${data.faceColor}" />"""
        }
        val cornerRadius = if (data.isCircle) SIZE else SIZE / 6
        // The library emits `  ${title ? "<title>…</title>" : ""}` on its own
        // line; the web never asks for a title, so the line is two spaces.
        val title = ""

        return """<svg
  viewBox="0 0 $SIZE $SIZE"
  fill="none"
  role="img"
  xmlns="http://www.w3.org/2000/svg"
  width="${size}px"
  height="${size}px"
>
  $title
  <mask id="$id" maskUnits="userSpaceOnUse" x="0" y="0" width="$SIZE" height="$SIZE">
    <rect width="$SIZE" height="$SIZE" rx="${SIZE * 2}" fill="#FFFFFF" />
  </mask>
  <g mask="url(#$id)">
    <rect width="$SIZE" height="$SIZE" fill="${data.backgroundColor}" />
    <rect
      x="0"
      y="0"
      width="$SIZE"
      height="$SIZE"
      transform="translate(${data.wrapperTranslateX} ${data.wrapperTranslateY}) rotate(${data.wrapperRotate} $half $half) scale(${scaleText(data.wrapperScaleTenths)})"
      fill="${data.wrapperColor}"
      rx="$cornerRadius"
    />
    <g transform="translate(${numberText(data.faceTranslateX)} ${numberText(data.faceTranslateY)}) rotate(${data.faceRotate} $half $half)">
      $mouth
      <rect x="${14 - data.eyeSpread}" y="14" width="1.5" height="2" rx="1" stroke="none" fill="${data.faceColor}" />
      <rect x="${20 + data.eyeSpread}" y="14" width="1.5" height="2" rx="1" stroke="none" fill="${data.faceColor}" />
    </g>
  </g>
</svg>"""
    }
}
